### PR TITLE
Fix issue 145 and 172

### PR DESCRIPTION
**Pull request by LeeiFrankJaw**
This squashed commit basically fix issue [145][1] and 3 points mentioned in issue [172][2].  It now supports basic UTF-8 characters and perform proper case conversion for common latin letters with diacritics.  At the same time, it is backward compatible with most existing syntactics. Nothing can demonstrate this better than the following examples.

```
"200 \LaTeX \ae Foö{bar}{\'o \ae}{{\'o}}" smart.upper.case top$
"Perturbations in the A $\Sigma_u^+$ state of Na$_2$" smart.sentence.case top$
```

The above code produce `200 \LaTeX \AE FOÖ{bar}{\'O \AE}{{\'o}}` and `Perturbations in the A $\Sigma_u^+$ state of Na$_2$. {\H {c}a{d{e}}}o`. It mostly respects `x_change_case` procedure implemented in [_bibtex.web_][3].  One obvious difference is that commands (except those for single letters with diacritics) at brace level 0 won't undergo case transformation.  Brace protection is still honored.  For UTF-8 characters and simple math expression, you probably won't need them though, as indicated in the above example.

[1]: https://github.com/zepinglee/gbt7714-bibtex-style/issues/145
[2]: https://github.com/zepinglee/gbt7714-bibtex-style/issues/172#issuecomment-2494493432
[3]: https://tug.org/svn/texlive/trunk/Build/source/texk/web2c/bibtex.web?revision=57915&view=markup#l8884

Squashed commit of the following:

commit bc1574fd7837046a3749f7679a030bd1c47d19e0
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Mon Dec 2 10:46:11 2024 +0800

    Rewrite normalize.page.range again

commit b54d85655923f7fc119c778aca8297e84ea9ba9c
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Fri Nov 29 03:31:35 2024 +0800

    Rewrite normalize.page.range

commit bdb9fdd71388f3f03365d6bc70b3213b96ae148d
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Thu Nov 28 08:22:00 2024 +0800

    Implement functions for texchar semantics

commit a1d3f5b5feb9ae88e33919dd706e45a79fff178e
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Wed Nov 27 17:26:57 2024 +0800

    Support Latin Extended-A

commit 775f79716f553066983f1b7e4afef259b0335133
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Wed Nov 27 16:33:01 2024 +0800

    Implement smart.upper.case

commit 20bffbdb3d111f4f54f3769e7fe56a2c6d1959bb
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Sat Nov 23 00:46:39 2024 +0800

    Support polymorphism when tokenizing

commit 31ef32edcf181f070ebd7159377d2e3f52690952
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Thu Nov 21 21:57:27 2024 +0800

    Update tests for Dublin Core entry

commit 4abe6018a40192cbf8dceaa7f5a86f682aaeb279
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Thu Nov 21 09:08:03 2024 +0800

    Increase compatibility of font selection

commit aad1cb912de60169f6bb11c5d8dfbbd210545129
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Thu Nov 21 06:48:52 2024 +0800

    Use page.range.separator

    Rewrite `hyphenate` and rename it to `normalize.page.range`.  Use
    `page.range.separator` to configure separator in page ranges.  Also
    update DocStrip options for added or modifed configuration variables.

commit 0450382a41a47269646aaa9ad22f81d43125afad
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Thu Nov 21 04:49:56 2024 +0800

    Follow the existing convention

    Follow the convention of using function to define constants

commit cf3c5c662e8bd125fc5e3d41ca12a6de507c0f4a
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Thu Nov 21 02:06:25 2024 +0800

    Refactoring

     * Now `is.all.lower` returns true for empty strings.  Following the convention of modern predicate logic, it assumes no existential import.  Update functions which depend on it.

     * The second argument of the return value of `split.first.char.from.str`
       is of polymorphic type.  It returns an empty string for an empty
       string instead of a null char.

     * Some functions are rewritten to enable short-circuit evaluation.

commit 352b89ba29034ffe8b1a6d45beddba7365215de3
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Wed Nov 20 19:30:07 2024 +0800

    Do some refactoring and renaming

commit bbf5add59a0cf7b620d737f87081f01830bfbceb
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Wed Nov 20 07:57:18 2024 +0800

    Enable lowercase.word.after.colon by default

    Also update tests for this

commit 6a7c3bc1b4ceafd3a0a14fe480e9a32a3445d01c
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Wed Nov 20 07:49:36 2024 +0800

    Update tests for smart.sentence.case

commit b3c9d79617e7f0fa761964872dddbc6b2fca7779
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Wed Nov 20 06:40:38 2024 +0800

    Add basic UTF-8 support

commit 456687e162eee4c480b9bd9adac31de964ae912a
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Sat Nov 16 05:53:13 2024 +0800

    Remove ignore.extra.interword.space

    This feature is extraneous since the extra spaces are already
    preprocessed by the BibTeX.

commit b15d6737f82ae87695a998fc7cd7bf2035b50683
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Thu Nov 14 20:33:20 2024 +0800

    Improve smart.sentence.case.lower.token

commit d7e7f532b6489556425275c57a0fd3bbfbecc330
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Thu Nov 14 19:18:29 2024 +0800

    Basically finish the smart lowercase feature

commit 5da7b113abbae12370ef1ebe1a1f5d347c617e6c
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Mon Nov 11 22:22:24 2024 +0800

    Update the source dtx file

commit dc88ba834c43ea88e48c04cf2e8c2dd959a1af23
Author: Lei Zhao <LeiZh26@gmail.com>
Date:   Mon Nov 11 20:38:39 2024 +0800

    Add en.dash.in.pages option

    Also process UTF-8 en dash (–)

### CODE REVIEW

**Comment by zepinglee** [comment]
感谢！不过我可能需要时间检查下有没有别的问题。

File: gbt7714.dtx [hunk 1]
```diff
@@ -1565,6 +1588,14 @@ FUNCTION {bbl.sine.loco.sine.nomine}
   if$
 }
 
+FUNCTION {default.self.tokens} { ":,-'–—?.!" }
+
+FUNCTION {latin.upper} { "ÀÁÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖØÙÚÛÜÝÞŸĀĂĄĆĈĊČĎĐĒĔĖĘĚĜĞĠĢĤĦĨĪĬĮİIĴĶĹĻĽĿŁŃŅŇŊŌŎŐŒŔŖŘŚŜŞŠŢŤŦŨŪŬŮŰŲŴŶŸŹŻŽ" }
+
+FUNCTION {latin.lower} { "àáãäåæçèéêëìíîïðñòóôõöøùúûüýþÿāăąćĉċčďđēĕėęěĝğġģĥħĩīĭįiıĵķĺļľŀłńņňŋōŏőœŕŗřśŝşšţťŧũūŭůűųŵŷÿźżž" }
+
+FUNCTION {range.separators} { "-–—~～" }
```
Comment (zepinglee) [inline gbt7714.dtx:1597]: 半角 `~` 不应视作页码范围的分隔符，因为不符合习惯。

**Comment by zepinglee** [comment]
另外我将 `page.range.separator` 的名称改成了 `page.range.delimiter`。

感谢 PR！
